File: ios/DisbandiOS/Services/NotesService.swift
import Foundation
import Observation
import Realtime
import Supabase

/// A private note. Owner-only by RLS (`auth.uid() = user_id`) — there is no
/// sharing path, so nothing here is ever visible to another account.
struct Note: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let userId: String
    var content: String
    var attachmentUrl: String?
    var attachmentType: AttachmentType?
    var attachmentName: String?
    var pinned: Bool
    let createdAt: String?
    var editedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, content, pinned
        case userId = "user_id"
        case attachmentUrl = "attachment_url"
        case attachmentType = "attachment_type"
        case attachmentName = "attachment_name"
        case createdAt = "created_at"
        case editedAt = "edited_at"
    }
}

private struct NewNote: Encodable {
    let userId: String
    let content: String
    let attachmentUrl: String?
    let attachmentType: String?
    let attachmentName: String?

    enum CodingKeys: String, CodingKey {
        case content
        case userId = "user_id"
        case attachmentUrl = "attachment_url"
        case attachmentType = "attachment_type"
        case attachmentName = "attachment_name"
    }
}

private struct NoteEdit: Encodable {
    let content: String
    let editedAt: String
    enum CodingKeys: String, CodingKey {
        case content
        case editedAt = "edited_at"
    }
}

private struct NotePin: Encodable {
    let pinned: Bool
}

/// Backs the Notes tab: paged history, pin/edit/delete, and realtime so a note
/// written on the desktop shows up here without a refresh.
@MainActor
@Observable
final class NotesService {
    private(set) var notes: [Note] = []
    private(set) var loading = true
    private(set) var hasMore = false
    var error: String?

    private static let pageSize = 50

    private var hasLoadedOnce = false
    private var channel: RealtimeChannelV2?
    private var watchTask: Task<Void, Never>?
    private var changeChannel: RealtimeChannelV2?
    private var changeTask: Task<Void, Never>?
    private var userId: String?

    private var client: SupabaseClient { SupabaseManager.client }

    var pinned: [Note] { notes.filter(\.pinned) }

    func start(userId: String?) async {
        guard self.userId != userId else { return }
        stop()
        self.userId = userId
        hasLoadedOnce = false
        notes = []
        guard userId != nil else {
            loading = false
            return
        }
        await load()
        await subscribe()
    }

    func stop() {
        watchTask?.cancel()
        watchTask = nil
        changeTask?.cancel()
        changeTask = nil
        let ch = channel
        let changes = changeChannel
        channel = nil
        changeChannel = nil
        Task {
            await ch?.unsubscribe()
            await changes?.unsubscribe()
        }
    }

    func load() async {
        guard let userId else { return }
        if !hasLoadedOnce { loading = true }
        defer {
            loading = false
            hasLoadedOnce = true
        }
        do {
            let rows: [Note] = try await client
                .from("notes")
                .select("*")
                .eq("user_id", value: userId)
                .order("created_at", ascending: false)
                .limit(Self.pageSize)
                .execute().value
            notes = rows
            hasMore = rows.count == Self.pageSize
        } catch {
            self.error = friendlyMessage(error)
        }
    }

    func loadMore() async {
        guard let userId, hasMore, let oldest = notes.last?.createdAt else { return }
        do {
            let rows: [Note] = try await client
                .from("notes")
                .select("*")
                .eq("user_id", value: userId)
                .lt("created_at", value: oldest)
                .order("created_at", ascending: false)
                .limit(Self.pageSize)
                .execute().value
            // Realtime may have already inserted one of these.
            let known = Set(notes.map(\.id))
            notes.append(contentsOf: rows.filter { !known.contains($0.id) })
            hasMore = rows.count == Self.pageSize
        } catch {
            self.error = friendlyMessage(error)
        }
    }

    func send(content: String, attachment: OutgoingAttachment? = nil) async {
        guard let userId else { return }
        let trimmed = content.trimmingCharacters(in: .whitespacesAndNewlines)
        // Matches the `notes_not_empty` check constraint.
        guard !trimmed.isEmpty || attachment != nil else { return }
        do {
            let payload = NewNote(
                userId: userId,
                content: trimmed,
                attachmentUrl: attachment?.url,
                attachmentType: attachment?.type,
                attachmentName: nil
            )
            let inserted: [Note] = try await client
                .from("notes").insert(payload).select("*").execute().value
            if let note = inserted.first, !notes.contains(where: { $0.id == note.id }) {
                notes.insert(note, at: 0)
            }
        } catch {
            self.error = friendlyMessage(error)
        }
    }

    func edit(_ note: Note, content: String) async {
        let trimmed = content.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty || note.attachmentUrl != nil else { return }
        let stamp = ISO8601DateFormatter().string(from: Date())
        do {
            try await client.from("notes")
                .update(NoteEdit(content: trimmed, editedAt: stamp))
                .eq("id", value: note.id)
                .execute()
            if let idx = notes.firstIndex(where: { $0.id == note.id }) {
                notes[idx].content = trimmed
                notes[idx].editedAt = stamp
            }
        } catch {
            self.error = friendlyMessage(error)
        }
    }

    func togglePin(_ note: Note) async {
        let next = !note.pinned
        // Optimistic: pinning should feel instant.
        if let idx = notes.firstIndex(where: { $0.id == note.id }) {
            notes[idx].pinned = next
        }
        do {
            try await client.from("notes")
                .update(NotePin(pinned: next))
                .eq("id", value: note.id)
                .execute()
        } catch {
            if let idx = notes.firstIndex(where: { $0.id == note.id }) {
                notes[idx].pinned = !next
            }
            self.error = friendlyMessage(error)
        }
    }

    func delete(_ note: Note) async {
        let previous = notes
        notes.removeAll { $0.id == note.id }
        do {
            try await client.from("notes").delete().eq("id", value: note.id).execute()
        } catch {
            notes = previous
            self.error = friendlyMessage(error)
        }
    }

    private func subscribe() async {
        guard let userId else { return }
        let (channel, stream) = await RealtimeService.observeInserts(
            table: "notes",
            filter: "user_id=eq.\(userId)",
            as: Note.self
        )
        self.channel = channel
        watchTask = Task { [weak self] in
            for await note in stream {
                await MainActor.run {
                    guard let self, !self.notes.contains(where: { $0.id == note.id }) else { return }
                    self.notes.insert(note, at: 0)
                }
            }
        }

        // Inserts alone are not enough: deleting, editing or pinning a note on
        // the desktop produced no event here, so the phone kept showing notes
        // that no longer existed. Any change triggers a reconcile against the
        // server.
        let (anyChannel, anyStream) = await RealtimeService.observeChanges(
            table: "notes",
            filter: "user_id=eq.\(userId)"
        )
        changeChannel = anyChannel
        changeTask = Task { [weak self] in
            for await _ in anyStream {
                await self?.reconcile()
            }
        }
    }

    /// Re-read the newest page and bring local state in line with it.
    ///
    /// Deletions are only applied within the window the fetch actually covers —
    /// an older note that simply falls outside the page must not be mistaken
    /// for one that was deleted.
    private func reconcile() async {
        guard let userId else { return }
        do {
            let rows: [Note] = try await client
                .from("notes")
                .select("*")
                .eq("user_id", value: userId)
                .order("created_at", ascending: false)
                .limit(Self.pageSize)
                .execute().value

            let serverIds = Set(rows.map(\.id))
            let windowFloor = rows.count == Self.pageSize ? rows.last?.createdAt : nil

            var merged = notes.filter { note in
                if serverIds.contains(note.id) { return true }
                // Outside the fetched window — keep it, we cannot tell.
                if let floor = windowFloor, let created = note.createdAt, created < floor {
                    return true
                }
                return false
            }

            // Apply server values for rows we still hold, and add anything new.
            for row in rows {
                if let idx = merged.firstIndex(where: { $0.id == row.id }) {
                    merged[idx] = row
                } else {
                    merged.append(row)
                }
            }
            merged.sort { ($0.createdAt ?? "") > ($1.createdAt ?? "") }
            notes = merged
        } catch {
            // A failed reconcile should not surface as a user-facing error;
            // the next change (or a manual refresh) will try again.
        }
    }

    private func friendlyMessage(_ error: Error) -> String {
        let text = error.localizedDescription
        if text.localizedCaseInsensitiveContains("notes") && text.localizedCaseInsensitiveContains("exist") {
            return "Notes isn't set up on this account yet."
        }
        return text
    }
}

File: android/app/src/main/java/com/wsgpolar/disband/data/NotesService.kt
package com.wsgpolar.disband.data

import com.wsgpolar.disband.core.TimeFormat
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Backs the Notes tab: paged history, pin/edit/delete, and realtime so a note
 * written on the desktop shows up here without a refresh. Mirrors the iOS
 * `NotesService`.
 */
class NotesService(private val scope: CoroutineScope) {
    private val _notes = MutableStateFlow<List<Note>>(emptyList())
    val notes: StateFlow<List<Note>> = _notes.asStateFlow()

    private val _loading = MutableStateFlow(true)
    val loading: StateFlow<Boolean> = _loading.asStateFlow()

    private val _hasMore = MutableStateFlow(false)
    val hasMore: StateFlow<Boolean> = _hasMore.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private var hasLoadedOnce = false
    private var watchJob: Job? = null
    private var changeJob: Job? = null
    private var userId: String? = null

    suspend fun start(userId: String?) {
        if (this.userId == userId) return
        stop()
        this.userId = userId
        hasLoadedOnce = false
        _notes.value = emptyList()
        if (userId == null) {
            _loading.value = false
            return
        }
        load()
        subscribe(userId)
    }

    fun stop() {
        watchJob?.cancel()
        watchJob = null
        changeJob?.cancel()
        changeJob = null
    }

    suspend fun load() {
        val uid = userId ?: return
        if (!hasLoadedOnce) _loading.value = true
        _loading.value = false
        try {
            val rows = Database.fetchNotes(uid, limit = 50)
            _notes.value = rows
            _hasMore.value = rows.size == 50
            hasLoadedOnce = true
        } catch (e: Exception) {
            _error.value = friendlyMessage(e)
        }
    }

    suspend fun loadMore() {
        val uid = userId ?: return
        if (!_hasMore.value) return
        val oldest = _notes.value.lastOrNull()?.createdAt ?: return
        try {
            val rows = Database.fetchNotes(uid, limit = 50, before = oldest)
            val known = _notes.value.map { it.id }.toSet()
            _notes.value = _notes.value + rows.filter { it.id !in known }
            _hasMore.value = rows.size == 50
        } catch (e: Exception) {
            _error.value = friendlyMessage(e)
        }
    }

    suspend fun send(content: String, attachment: OutgoingAttachment? = null) {
        val uid = userId ?: return
        val trimmed = content.trim()
        if (trimmed.isEmpty() && attachment == null) return
        try {
            val note = Database.insertNote(uid, trimmed, attachment) ?: return
            if (_notes.value.none { it.id == note.id }) {
                _notes.value = listOf(note) + _notes.value
            }
        } catch (e: Exception) {
            _error.value = friendlyMessage(e)
        }
    }

    suspend fun edit(note: Note, content: String) {
        val trimmed = content.trim()
        if (trimmed.isEmpty() && note.attachmentUrl == null) return
        val stamp = TimeFormat.nowStamp()
        try {
            Database.updateNoteContent(note.id, trimmed, stamp)
            _notes.value = _notes.value.map {
                if (it.id == note.id) it.copy(content = trimmed, editedAt = stamp) else it
            }
        } catch (e: Exception) {
            _error.value = friendlyMessage(e)
        }
    }

    suspend fun togglePin(note: Note) {
        val next = !note.pinned
        _notes.value = _notes.value.map { if (it.id == note.id) it.copy(pinned = next) else it }
        try {
            Database.setNotePinned(note.id, next)
        } catch (e: Exception) {
            _notes.value = _notes.value.map { if (it.id == note.id) it.copy(pinned = !next) else it }
            _error.value = friendlyMessage(e)
        }
    }

    suspend fun delete(note: Note) {
        val previous = _notes.value
        _notes.value = previous.filterNot { it.id == note.id }
        try {
            Database.deleteNote(note.id)
        } catch (e: Exception) {
            _notes.value = previous
            _error.value = friendlyMessage(e)
        }
    }

    fun pinned(): List<Note> = _notes.value.filter { it.pinned }

    private fun subscribe(userId: String) {
        scope.launch {
            try {
                val live = RealtimeService.observeInserts("notes", "user_id=eq.$userId", Note.serializer())
                watchJob = launch {
                    live.flow.collect { note ->
                        if (_notes.value.none { it.id == note.id }) {
                            _notes.value = listOf(note) + _notes.value
                        }
                    }
                }
            } catch (_: Exception) {
            }
        }

        scope.launch {
            try {
                val changes = RealtimeService.observeChanges("notes", "user_id=eq.$userId")
                changeJob = launch {
                    changes.flow.collect { reconcile() }
                }
            } catch (_: Exception) {
            }
        }
    }

    /** Re-read the newest page and bring local state in line with it. */
    private suspend fun reconcile() {
        val uid = userId ?: return
        try {
            val rows = Database.fetchNotes(uid, limit = 50)
            val serverIds = rows.map { it.id }.toSet()
            val windowFloor = if (rows.size == 50) rows.lastOrNull()?.createdAt else null

            val merged = _notes.value.filter { note ->
                if (note.id in serverIds) return@filter true
                val created = note.createdAt ?: return@filter false
                if (windowFloor != null && created < windowFloor) return@filter true
                false
            }.toMutableList()

            for (row in rows) {
                val idx = merged.indexOfFirst { it.id == row.id }
                if (idx >= 0) merged[idx] = row else merged.add(row)
            }
            merged.sortByDescending { it.createdAt.orEmpty() }
            _notes.value = merged
        } catch (_: Exception) {
        }
    }

    private fun friendlyMessage(e: Exception): String {
        val text = e.message ?: e.toString()
        return if ("notes" in text && "exist" in text) {
            "Notes isn't set up on this account yet."
        } else {
            text
        }
    }
}